codegen/smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/smoketests/SmokeTestsRunnerGenerator.kt

```diff
@@ -0,0 +1,216 @@
+package software.amazon.smithy.kotlin.codegen.rendering.smoketests
+
+import software.amazon.smithy.codegen.core.Symbol
+import software.amazon.smithy.kotlin.codegen.core.*
+import software.amazon.smithy.kotlin.codegen.integration.SectionId
+import software.amazon.smithy.kotlin.codegen.model.expectShape
+import software.amazon.smithy.kotlin.codegen.model.getTrait
+import software.amazon.smithy.kotlin.codegen.model.hasTrait
+import software.amazon.smithy.kotlin.codegen.model.traits.FailedResponseTrait
+import software.amazon.smithy.kotlin.codegen.model.traits.SuccessResponseTrait
+import software.amazon.smithy.kotlin.codegen.rendering.util.format
+import software.amazon.smithy.kotlin.codegen.utils.dq
+import software.amazon.smithy.kotlin.codegen.utils.operations
+import software.amazon.smithy.kotlin.codegen.utils.toCamelCase
+import software.amazon.smithy.model.shapes.OperationShape
+import software.amazon.smithy.model.shapes.ServiceShape
+import software.amazon.smithy.smoketests.traits.SmokeTestCase
+import software.amazon.smithy.smoketests.traits.SmokeTestsTrait
+import kotlin.jvm.optionals.getOrNull
+
+object SmokeTestsRunner : SectionId
+
+// Env var constants
+const val SKIP_TAGS = "AWS_SMOKE_TEST_SKIP_TAGS"
+const val SERVICE_FILTER = "AWS_SMOKE_TEST_SERVICE_IDS"
+
+/**
+ * Renders smoke tests runner for a service
+ */
+class SmokeTestsRunnerGenerator(
+    private val writer: KotlinWriter,
+    ctx: CodegenContext,
+) {
+    // Generator config
+    private val model = ctx.model
+    private val sdkId = ctx.settings.sdkId
+    private val symbolProvider = ctx.symbolProvider
+    private val service = symbolProvider.toSymbol(model.expectShape(ctx.settings.service))
+    private val operations = ctx.model.operations(ctx.settings.service).filter { it.hasTrait<SmokeTestsTrait>() }
+
+    // Test config
+    private val hasSuccessResponseTrait = ctx.model.expectShape<ServiceShape>(ctx.settings.service).hasTrait(SuccessResponseTrait.ID)
+    private val hasFailedResponseTrait = ctx.model.expectShape<ServiceShape>(ctx.settings.service).hasTrait(FailedResponseTrait.ID)
+    init {
+        check(!(hasSuccessResponseTrait && hasFailedResponseTrait)) {
+            "A service can't have both the success response trait and the failed response trait."
+        }
+    }
```

Comment:
Do we also need to check that `hasSuccessResponseTrait || hasFailedResponseTrait` (a service has exactly one of the traits)?

Reply:
No because the trait is only used for the E2E tests. During code generation for actual real life services neither of the traits is necessary

Reply:
Yeah, I don't like it either. I can't think of any other reasonable way to replace the `httpClient` without high coupling. 

> If we're just using these traits to configure the httpClient, can it be configured through the smoke test vendor params instead?

We would have to add some custom logic for that to work, so the coupling would still be there.